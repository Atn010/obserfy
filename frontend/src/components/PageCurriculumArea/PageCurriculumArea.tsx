import React, { FC } from "react"
import useApi from "../../api/useApi"
import Box from "../Box/Box"
import Typography from "../Typography/Typography"
import LoadingPlaceholder from "../LoadingPlaceholder/LoadingPlaceholder"
import Card from "../Card/Card"
import BackNavigation from "../BackNavigation/BackNavigation"

// FIXME: Typescript any typing, and inconsistent loading state should be fixed.
interface Props {
  id: string
}
interface Area {
  id: string
  name: string
}
interface Subject {
  id: string
  name: string
  order: string
}
export const PageCurriculumArea: FC<Props> = ({ id }) => {
  const [area, setAreaOutdated, areaLoading] = useApi<Area>(
    `/curriculum/areas/${id}`
  )
  const [subjects, setSubjectsOutdated, subjectsLoading] = useApi<Subject[]>(
    `/curriculum/areas/${id}/subjects`
  )
  const loading = areaLoading || subjectsLoading

  const subjectList = subjects?.map(subject => (
    <Box m={3} key={subject.id}>
      <SubjectMaterials subject={subject} />
    </Box>
  ))

  return (
    <Box maxWidth="maxWidth.sm" margin="auto">
      <BackNavigation to="/dashboard/settings/curriculum" text="Curriculum" />
      {loading && <LoadingState />}
      <Typography.H3 p={3} lineHeight={1}>
        {area?.name}
      </Typography.H3>
      {!loading && subjectList}
    </Box>
  )
}

const LoadingState: FC = () => (
  <Box p={3}>
    <LoadingPlaceholder width="100%" height="4rem" />
  </Box>
)

interface SubjectProps {
  subject: Subject
}
interface Material {
  id: string
  name: string
  order: number
}
const SubjectMaterials: FC<SubjectProps> = ({ subject }) => {
  const [materials, setMaterialsOutdated, loading] = useApi<Material[]>(
    `/curriculum/subjects/${subject.id}/materials`
  )

  const materialList = materials?.map(material => (
    <Box mx={3} my={2} key={material.id}>
      <Typography.Body>{material.name}</Typography.Body>
    </Box>
  ))

  const loadingPlaceholder = loading && (
    <Box m={3}>
      <LoadingPlaceholder width="100%" height="4rem" mb={3} />
      <LoadingPlaceholder width="100%" height="4rem" mb={3} />
      <LoadingPlaceholder width="100%" height="4rem" mb={3} />
      <LoadingPlaceholder width="100%" height="4rem" mb={3} />
      <LoadingPlaceholder width="100%" height="4rem" mb={3} />
    </Box>
  )

  return (
    <Card mb={3} pb={2}>
      <Typography.H5 m={3} mb={4}>
        {subject.name}
      </Typography.H5>
      {loadingPlaceholder}
      {materialList}
    </Card>
  )
}

export default PageCurriculumArea
