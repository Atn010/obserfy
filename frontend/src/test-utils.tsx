/* eslint-disable import/no-extraneous-dependencies */
import React, { FC, ReactElement } from "react"
import { ThemeProvider } from "theme-ui"
import { render, RenderOptions, RenderResult } from "@testing-library/react"
import { IntlProvider } from "gatsby-plugin-intl3"
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router"
import Theme from "./gatsby-plugin-theme-ui"
import Box from "./components/Box/Box"
import en from "./intl/en.json"

const AllProviders: FC = ({ children }) => {
  // TODO: Maybe this better be outside
  const history = createHistory(createMemorySource("/"))
  return (
    <LocationProvider history={history}>
      <ThemeProvider theme={Theme}>
        <IntlProvider locale="en" messages={en}>
          <Box fontSize={[16, 20]}>{children}</Box>
        </IntlProvider>
      </ThemeProvider>
    </LocationProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: AllProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
