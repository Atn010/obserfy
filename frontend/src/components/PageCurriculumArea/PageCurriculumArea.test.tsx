import React from "react"
import { render } from "../../test-utils"
import { Basic } from "./PageCurriculumArea.stories"

describe("PageCurriculumArea", () => {
  it("should render correctly", () => {
    const { container } = render(<Basic />)
    expect(container).toMatchSnapshot()
  })
})
