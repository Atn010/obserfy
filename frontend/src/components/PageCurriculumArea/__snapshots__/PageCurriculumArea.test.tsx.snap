// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageCurriculumArea should render correctly 1`] = `
@keyframes animation-0 {
  0% {
    background-color: rgba(165,165,165,0.05);
  }

  50% {
    background-color: rgba(165,165,165,0.1);
  }

  100% {
    background-color: rgba(165,165,165,0.05);
  }
}

.emotion-7 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 16px;
}

@media screen and (min-width:40em) {
  .emotion-7 {
    font-size: 20px;
  }
}

.emotion-6 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 640px;
  margin: auto;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: -8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  max-width: 100%;
  height: auto;
  fill: var(--theme-ui-colors-icon,rgba(0,0,0,0.6));
  margin: 16px;
  height: 24px;
  width: 24px;
  margin-right: 4px;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: rgba(0,0,0,0.88);
  font-size: 1.6rem;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  line-height: 3rem;
  color: var(--theme-ui-colors-textMediumEmphasis,rgba(0,0,0,0.60));
  font-size: 16px;
  margin-bottom: 0;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: 14px;
  }
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding: 16px;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  border-radius: 6px;
  -webkit-animation: 1s ease-in-out 0s infinite animation-0;
  animation: 1s ease-in-out 0s infinite animation-0;
  width: 100%;
  height: 4rem;
}

.emotion-5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 24px;
  font-family: Open Sans,-apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 300;
  line-height: heading;
  font-size: 3.90625rem;
  line-height: 6rem;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: var(--theme-ui-colors-text,rgba(0,0,0,0.87));
  padding: 16px;
  line-height: 1;
}

<div>
  <div
    class="emotion-7"
  >
    <div
      class="emotion-6"
    >
      <a
        href="/dashboard/settings/curriculum"
      >
        <div
          class="emotion-2"
        >
          <div
            class="emotion-0"
          />
          <p
            class="emotion-1"
          >
            Curriculum
          </p>
        </div>
      </a>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-3"
        />
      </div>
      <h3
        class="emotion-5"
      />
    </div>
  </div>
</div>
`;
